lib/phoenix/sync/write.ex

```diff
@@ -0,0 +1,970 @@
+                    allow for mapping between table names on the client and within Postgres.
+                    """
+                  ],
+                  load: [
+                    type: {:or, [{:fun, 1}, {:fun, 2}, :mfa]},
+                    doc: """
+                    A 1- or 2-arity function that accepts either the
+                    mutation data or an `Ecto.Repo` instance and the mutation data and returns
+                    the original row from the database or `nil` if not found.
+
+                    This function is only used for updates or deletes. For
+                    inserts, the `__struct__/0` function defined by `Ecto.Schema` is used to
+                    create an empty schema struct.
+
+                    ## Examples
+
+                        # load from a known Repo
+                        load: fn %{"id" => id} -> MyApp.Repo.get(Todos.Todo, id)
+
+                        # load from the repo passed to `#{__MODULE__}.transaction/2`
+                        load: fn repo, %{"id" => id} -> repo.get(Todos.Todo, id)
+
+                    If not provided defaults to `c:Ecto.Repo.get_by/3` using the
+                    table's schema module and its primary keys.
+                    """,
+                    type_spec:
+                      quote(
+                        do:
+                          (Ecto.Repo.t(), data() -> Ecto.Schema.t() | nil)
+                          | (data() -> Ecto.Schema.t() | nil)
+                      )
+                  ],
+                  accept: [
+                    type: {:list, {:in, @actions}},
+                    doc: """
+                    A list of actions to accept. Defaults to accepting all operations, `#{inspect(@actions)}`.
+                    """
+                  ],
+                  changeset: [
+                    type: {:or, [{:fun, 3}, :mfa]},
+                    doc: """
+                    `changeset` should be a 3-arity function that returns
+                    an `Ecto.Changeset` for a given mutation.
+
+                    ### Callback params
+
+                    - `data` an Ecto.Schema struct matching the one used when
+                      calling `allow/2` returned from the `load` function.
+                    - `action` the operation action, one of `:insert`, `:update` or `:delete`
+                    - `changes` a map of changes to apply to the `data`.
+
+                    At absolute minimum, this should call
+                    `Ecto.Changeset.cast/3` to validate the proposed data:
+
+                        def my_changeset(data, _action, changes) do
+                          Ecto.Changeset.cast(data, changes, @permitted_columns)
+                        end
+
+                    But it's vitally important to remember that the `data` and
+                    `changes` are **untrusted** values.
+
+                    Unlike standard uses of `Ecto.Changeset` functions, where
```

Comment:
Could we separate out the auth layer in this stack then? I.e.: have a `auth` callback and a `changes` callback, so that the data can be authorized before passing to the standard changes function?

Reply:
Rather than have two functions, maybe just rename `changeset` to something else, like `validate` or something. Within that we can call the standard `changeset/2` function and then do the auth[n,z] of the resulting changeset.